src/tests_exhaustive.c

```diff
@@ -120,26 +155,90 @@ void test_exhaustive_addition(const secp256k1_ge *group, const secp256k1_gej *gr
     }
 }
 
-void test_exhaustive_ecmult(secp256k1_context *ctx, secp256k1_ge *group, secp256k1_gej *groupj, int order) {
-    int i, j;
-    const int r_log = secp256k1_rand32() % order;  /* TODO be less biased */
-    for (j = 0; j < order; j++) {
-        for (i = 0; i < order; i++) {
-            secp256k1_gej tmp;
-            secp256k1_scalar na, ng;
-            secp256k1_scalar_set_int(&na, i);
-            secp256k1_scalar_set_int(&ng, j);
+void test_exhaustive_ecmult(const secp256k1_context *ctx, const secp256k1_ge *group, const secp256k1_gej *groupj, int order) {
+    int i, j, r_log;
+    for (r_log = 1; r_log < order; r_log++) {
+        for (j = 0; j < order; j++) {
+            for (i = 0; i < order; i++) {
+                secp256k1_gej tmp;
+                secp256k1_scalar na, ng;
+                secp256k1_scalar_set_int(&na, i);
+                secp256k1_scalar_set_int(&ng, j);
 
-            secp256k1_ecmult(&ctx->ecmult_ctx, &tmp, &groupj[r_log], &na, &ng);
-            ge_equals_gej(&group[(i * r_log + j) % order], &tmp);
+                secp256k1_ecmult(&ctx->ecmult_ctx, &tmp, &groupj[r_log], &na, &ng);
+                ge_equals_gej(&group[(i * r_log + j) % order], &tmp);
 
-            /* TODO we cannot exhaustively test ecmult_const as it does a scalar
-             * negation for even numbers, and our code is not designed to handle
-             * such a small scalar modulus. */
+                if (i > 0) {
+                    secp256k1_ecmult_const(&tmp, &group[i], &ng);
+                    ge_equals_gej(&group[(i * j) % order], &tmp);
+                }
+            }
         }
     }
 }
 
+void r_from_k(secp256k1_scalar *r, const secp256k1_ge *group, int k) {
+    secp256k1_fe x;
+    unsigned char x_bin[32];
+    k %= EXHAUSTIVE_TEST_ORDER;
+    x = group[k].x;
+    secp256k1_fe_normalize(&x);
+    secp256k1_fe_get_b32(x_bin, &x);
+    secp256k1_scalar_set_b32(r, x_bin, NULL);
+}
+
+/* hee hee hee */
+int solve_discrete_log(const secp256k1_scalar *x_coord, const secp256k1_ge *group, int order) {
+    int i;
+    for (i = 0; i < order; i++) {
+        secp256k1_scalar check_x;
+        r_from_k(&check_x, group, i);
+        if (*x_coord == check_x) {
+           return i;
+        }
+    }
+    return -1;
+}
+
+void test_exhaustive_sign(const secp256k1_context *ctx, const secp256k1_ge *group, int order) {
```

Comment:
Is there a use in passing order as a parameter? Almost all of this code already depends on knowing it is equal to EXHAUSTIVE_TEST_ORDER anyway.